index-evolution/src/test/java/fax/play/smoke/BasicsForNoDowntimeUpgradesTest.java

```diff
@@ -0,0 +1,340 @@
+        ModelUtils.createModel1Entities(cacheV2, 5, i -> ModelUtils.createModelD(2).apply(i + 10)); // Creating entities with ids +10 offset
+
+        // This is to check that older version is correctly writing also deprecated name field even though current
+        // schema present in the Infinispan server doesn't contain it
+        RemoteCache<String, Model3D> cacheV2Typed = cacheProviderV2.getCacheManager().getCache(CACHE1_NAME);
+        Model3D model3D = cacheV2Typed.get("300013");
+        assertThat(model3D.name).isEqualTo("modelD # 13");
+
+        // Query on nameIndexed should work with all entities >= 2 (e.g. all in the storage because we did the migration `migrateEntityAToEntityD`)
+        doQuery("FROM Model3 WHERE nameIndexed : '*3*'", cache, 4);
+    }
+
+    /**
+     * This example shows how an index can be removed
+     *
+     * In this example we assume the reason for removing the index is that it is no longer used in any query anymore
+     */
+    @Test
+    void testRemoveIndexWhenNoLongerUsedInQueryInNewerVersion() {
+        // VERSION 1
+        RemoteCache<String, Model> cache = cacheProvider
+                .init(new CacheDefinition(CACHE1_NAME, "Model3"), Schema3B.INSTANCE)
+                .getCache(CACHE1_NAME);
+
+        // Create VERSION 1 entities
+        ModelUtils.createModel1Entities(cache, 5, ModelUtils.createModelB(1));
+
+        // VERSION 1 uses name in a query
+        doQuery("FROM Model3 WHERE name LIKE '%3%'", cache, 1);
+
+        // VERSION 2
+        // In this version we cannot remove the index as node of VERSION 1 can still make a query with name in it
+        // The important part of this version is to remove all occurrences of the name field in queries
+        // Note: this version is needed only for analyzed fields, non-analyzed fields could work even without this,
+        // although currently we need it as well unless https://issues.redhat.com/browse/ISPN-8584 is resolved
+
+        // VERSION 3
+        // In this version we can be sure there is no query with name in it, therefore we can remove the index
+
+        // Update schema to not include index on name
+        // Note: If the reason for removal is removal of field completely, the process would be the same with the difference,
+        //  that this schema does not contain the field
+        cache = cacheProvider
+                .updateSchemaAndGet(Schema3A.INSTANCE)
+                .getCache(CACHE1_NAME);
+
+        // update index schema
+        cacheProvider.updateIndexSchema(CACHE1_NAME);
+
+        // Create entities without index
+        ModelUtils.createModel1Entities(cache, 5, ModelUtils.createModelA(2));
+
+        // Try query with field that has the index in both versions
+        doQuery("FROM Model3 WHERE entityVersion >= 1", cache, 10);
+        // TODO: Can we somehow cleanup indexed data only for name field?
```

Comment:
it would be a conditional reindexing or a conditional purge for this specific case.
you can open an issue to support this feature

Reply:
Created: https://issues.redhat.com/browse/ISPN-13948

Reply:
Thanks!